File: src/test/java/org/snakeyaml/engine/usecases/tags/LocalTagTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.usecases.tags;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.snakeyaml.engine.v2.api.ConstructNode;
import org.snakeyaml.engine.v2.api.Load;
import org.snakeyaml.engine.v2.api.LoadSettings;
import org.snakeyaml.engine.v2.nodes.Tag;

/**
 * Example of parsing a local tag
 */
@org.junit.jupiter.api.Tag("fast")
public class LocalTagTest {

  @Test
  public void testLocalTag() {
    Map<Tag, ConstructNode> tagConstructors = new HashMap<>();
    // register to call CustomConstructor when the Tag !ImportValue is found
    tagConstructors.put(new Tag("!ImportValue"), new CustomConstructor());
    LoadSettings settings = LoadSettings.builder().setTagConstructors(tagConstructors).build();
    Load loader = new Load(settings);
    Object obj = loader.loadFromString("VpcId: !ImportValue SpokeVPC");
    Map<String, ImportValueImpl> map = (Map<String, ImportValueImpl>) obj;
    assertEquals("SpokeVPC", map.get("VpcId").getValue());
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/usecases/tags/LocalTagTest.kt
@file:Suppress("UNCHECKED_CAST")
package it.krzeminski.snakeyaml.engine.kmp.usecases.tags

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import it.krzeminski.snakeyaml.engine.kmp.api.ConstructNode
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings
import it.krzeminski.snakeyaml.engine.kmp.nodes.Node
import it.krzeminski.snakeyaml.engine.kmp.nodes.ScalarNode
import it.krzeminski.snakeyaml.engine.kmp.nodes.Tag

/**
 * Example of parsing a local tag
 */
class LocalTagTest : FunSpec({
    test("local tag") {
        // register to call CustomConstructor when the Tag !ImportValue is found
        val tagConstructors = mapOf(Tag("!ImportValue") to CustomConstructor())
        val settings = LoadSettings.builder().setTagConstructors(tagConstructors).build()
        val loader = Load(settings)
        val obj = loader.loadOne("VpcId: !ImportValue SpokeVPC")
        val map = obj as Map<String, ImportValueImpl>
        map["VpcId"]!!.value shouldBe "SpokeVPC"
    }
})

/**
 * Create ImportValueImpl from a scalar node
 */
private class CustomConstructor : ConstructNode {
    override fun construct(node: Node?): Any {
        val scalar = node as ScalarNode
        return ImportValueImpl(scalar.value)
    }
}

/**
 * Business value to be parsed from YAML
 */
private class ImportValueImpl(val value: String)
